pkg/gator/reader/filereader.go

```diff
@@ -0,0 +1,223 @@
+package reader
+
+import (
+	"bufio"
+	"fmt"
+	"os"
+	"path/filepath"
+
+	"github.com/open-policy-agent/gatekeeper/pkg/oci"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+)
+
+var allowedExtensions = []string{".yaml", ".yml", ".json"}
+
+func ReadSources(filenames []string, images []string, tempDir string) ([]*unstructured.Unstructured, error) {
+	var sources []*source
+
+	// Read from --filename flag
+	s, err := readFiles(filenames)
+	if err != nil {
+		return nil, fmt.Errorf("reading from filenames: %w", err)
+	}
+	sources = append(sources, s...)
+
+	// Read from --image flag
+	s, err = readImages(images, tempDir)
+	if err != nil {
+		return nil, fmt.Errorf("pulling image: %w", err)
+	}
+	sources = append(sources, s...)
+
+	// Read stdin
+	stdinUnstructs, err := readStdin()
+	if err != nil {
+		return nil, fmt.Errorf("reading from stdin: %w", err)
+	}
+	sources = append(sources, &source{stdin: true, objs: stdinUnstructs})
+
+	conflicts := detectConflicts(sources)
+	for i := range conflicts {
+		logConflict(&conflicts[i])
+	}
+
+	return sourcesToUnstruct(sources), nil
+}
+
+func readImage(image string, tempDir string) ([]*source, error) {
+	dirPath, closeHandler, err := oci.PullImage(image, tempDir)
+	if closeHandler != nil {
+		defer closeHandler()
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	sources, err := readFile(dirPath)
+	if err != nil {
+		return nil, err
+	}
+	for _, s := range sources {
+		s.image = image
+	}
+
+	return sources, nil
+}
+
+func readImages(images []string, tempDir string) ([]*source, error) {
+	var sources []*source
+	for _, image := range images {
+		s, err := readImage(image, tempDir)
+		if err != nil {
+			return nil, err
+		}
+		sources = append(sources, s...)
+	}
+
+	return sources, nil
+}
+
+func readFile(filename string) ([]*source, error) {
+	if err := verifyFile(filename); err != nil {
+		return nil, err
+	}
+
+	var sources []*source
+	normalized, err := normalize([]string{filename})
```

Comment:
Can you pls give an example of a normalized filename? maybe also add a test for it?

Reply:
I think "normalize" in this case means to expand all files in the directory. Is that the case?

If so, would a less ambiguous name, like expandDirectories() avoid confusion?

Reply:
It's entirely possible that I'm missing the design here. I'm trying to understand when we need to expand files in the directory since it looks like we are passing in a directory for the images to be created. So when do we need to expand the files within that directory and why do we need to do that?

Reply:
IIUC the image needs to be extracted (essentially `tar -xvf`) in order to access the contents of the image.

It looks like the code is extracting the image contents to a temp dir, then walking that dir, extracting relevant YAML.

Mapping each file to the image is useful for error reporting. E.g. if a user has the same constraint with different values in two images, we're able to say which images are the source of the conflict (the goal being to avoid accidentally clobbering policy due to naming conflicts).

Reply:
Yep, Max is correct. 

It's also worth noting that this `normalize` business is not new code --it's the same code used by gator test|verify|expand, just moved into a new package `gator/reader`. This functionality is already tested in the gator e2e tests.

Reply:
Gotcha. Thanks for the explanations!